build_tools/aws-sdk-code-generator/templates/endpoints_plugin.mustache

```diff
@@ -44,11 +44,27 @@ module {{module_name}}
           context[:auth_scheme] =
             Aws::Endpoints.resolve_auth_scheme(context, endpoint)
 
-          @handler.call(context)
+          with_endpoint_metric(context.config) do
```

Comment:
I find these a little clunky.  Two thoughts on making it a little easier - 

1. For many metric cases (like these), its scoped within a single request.  We can add these metrics to the context instead of to a thread local.  That way, they do not need to a call a block and pop the metric off.   This both lets do conditional metrics more easily and avoids having to wrap the handler call, which isn't really necessary for these.  (Note: I'm suggesting a second way to set metrics - scoped only to the request.  Not replacing the thread local, which I know we need for other cases).

2.   If we do not want to support multiple places/scopes metrics can be added, then we can have `with_metric` take an array of metrics, and allow that array to be empty.  That will let us easily support conditional metrics and multiple ones.

Reply:
Yeah, it's pretty clunky especially as these add. I think we want to keep a single mechanism (thread local) but probably this can just take an array and we build the array and block call with it.